pandas/io/parsers/readers.py

```diff
@@ -941,10 +944,10 @@ def _get_options_with_defaults(self, engine):
 
     def _check_file_or_buffer(self, f, engine):
         # see gh-16530
-        if is_file_like(f) and engine != "c" and not hasattr(f, "__next__"):
-            # The C engine doesn't need the file-like to have the "__next__"
-            # attribute. However, the Python engine explicitly calls
-            # "__next__(...)" when iterating through such an object, meaning it
+        if is_file_like(f) and engine != "c" and not hasattr(f, "__iter__"):
```

Comment:
`__iter__` by itself seems to be enough.

Reply:
Previously `TextFileReader._check_file_or_buffer` would raise a `ValueError` for the "python" engine if `f` was a `tempfile.SpooledTemporaryFile` (which implements `__iter__` but **not** `__next__`)

Just wanted to check if `SpooledTemporaryFile` now works in the "python" engine for `TextFileReader` (and thus [this test should no longer skip the python parser](https://github.com/pandas-dev/pandas/blob/945c9ed766a61c7d2c0a7cbb251b6edebf9cb7d5/pandas/tests/io/parser/test_encoding.py#L261-L265)) or if this is an oversight? Sorry if this is a very niche case but it's something I've just run into on `1.3.4` and noticed that this logic had been recently changed in this PR

Reply:
`SpooledTemporaryFile` works if you open it in text mode, e.g.,  `SpooledTemporaryFile(mode="r+t")`.

The issue is that `SpooledTemporaryFile` does not have the attribute/property `readable` which `io.TextIOWrapper` requires (if you open `SpooledTemporaryFile` in binary mode). If you are lucky, you can convince the cpython maintainers to make `SpooledTemporaryFile` compatible with `io.TextIOWrapper`  in a future python version?

Reply:
Technically, we should add `readable` to the binary protocols that are wrapped in TextIOWrapper (I think only for read_json and read_read_csv).